core/src/main/scala/com/eed3si9n/jarjarabrams/Shader.scala

```diff
@@ -28,52 +29,74 @@ object Shader {
       mappings: Seq[(InputStream, String)],
       verbose: Boolean
   ): Seq[(InputStream, String)] = {
-    val jjrules = rules.flatMap { r =>
-      r.shadePattern match {
-        case ShadePattern.Rename(patterns) =>
-          patterns.map { case (from, to) =>
-            val jrule = new Rule()
-            jrule.setPattern(from)
-            jrule.setResult(to)
-            jrule
-          }
-        case ShadePattern.Zap(patterns) =>
-          patterns.map { pattern =>
-            val jrule = new Zap()
-            jrule.setPattern(pattern)
-            jrule
-          }
-        case ShadePattern.Keep(patterns) =>
-          patterns.map { pattern =>
-            val jrule = new Keep()
-            jrule.setPattern(pattern)
-            jrule
-          }
-        case _ => Nil
+    val shader = inputStreamShader(rules, verbose)
+    mappings.flatMap { case (inputStream, mapping) => shader(inputStream, mapping) }
+  }
+
+  def inputStreamShader(
+      rules: Seq[ShadeRule],
+      verbose: Boolean
+  ): (InputStream, String) => Option[(InputStream, String)] =
+    (inputStream, mapping) =>
+      if (rules.isEmpty) Some(inputStream -> mapping)
+      else {
+        bytcodeShader(rules, verbose)(readAllBytes(inputStream), mapping).map { case (bytes, newMapping) =>
+          new ByteArrayInputStream(bytes) { override def close(): Unit = inputStream.close() } -> newMapping
```

Comment:
Why close `inputStream` via this one? I'm not sure that makes sense, as `inputStream` is read before that (via `readAllBytes(inputStream)`).

Reply:
Don't we need to maintain the close chain? Otherwise we risk of leaving open `InputStream`.